twpca/crossval.py

```diff
@@ -0,0 +1,183 @@
+    """
+    idx = np.arange(N)
+    i = 0
+    while i < N-1:
+        r = range(i, i + n)
+        i += n
+        train_idx = np.setdiff1d(idx, r)
+        test_idx = idx[r]
+        yield train_idx, test_idx    
+
+def cross_validate(model, data, method, K, max_crossval=np.inf, **fit_kw):
+    """Runs specified cross-validation method.
+
+    Args:
+        model: TWPCA model instance
+        data: array-like (n_trials x n_timepoints x n_neurons)
+        method: string specifying method {'kfold', 'leaveout'}
+        K: int, crossvalidation parameter
+
+    Returns:
+        results: dict, model parameters and metrics calculated for all sessions
+    """
+
+    if method == 'kfold':
+        partitions = k_fold(data.shape[2], K)
+    elif method == 'leaveout':
+        partitions = leave_k_out(data.shape[2], K)
+    else:
+        raise ValueError('Cross-validation method not recognized.')
+
+    nfits = 0
+    results = []
+    for train, test in partitions:
+        # partition dataset
+        traindata = np.atleast_3d(data[:, :, train])
+        testdata = np.atleast_3d(data[:, :, test])
+
+        # fit model to training set
+        tf.reset_default_graph()
+        sess = tf.Session()
+        model.fit(traindata, sess=sess, **fit_kw)
+        # TODO - warmstart fits?
+
+        # assess dimensionality of warped vs unwarped testdata
+        warped_testdata = model.transform(testdata)
+        warped_rank, unwarped_rank = [], []
+        for n in range(testdata.shape[2]):
+            warped_rank.append(stable_rank(warped_testdata[:, :, n]))
+            unwarped_rank.append(stable_rank(testdata[:, :, n]))
+
+        # evaluate test error
+        time_factors = sess.run(model._warped_time_factors).reshape(-1, model.n_components)
+        _, resid, _, _ = np.linalg.lstsq(time_factors, testdata.reshape(-1, len(test)))
+
+        # compile results
+        results.append(
+            {'params': model.params,
+             'warped_rank': np.array(warped_rank),
+             'unwarped_rank': np.array(unwarped_rank),
+             'test_idx': test,
+             'train_idx': train,
+             'train_error': model._sess.run(model.recon_cost),
+             'test_error': np.sum(resid) / len(testdata.ravel()),
+             'obj_history': model.obj_history
+            }
+        )
+
+        # clean-up tensorflow
+        sess.close()
+
+        # terminate early if user is impatient
+        nfits += 1
+        if nfits >= max_crossval:
+            break
+
+    return results
+
+def hyperparam_search(data, n_components, warp_scales, time_scales,
+                      warp_reg=None, time_reg=None,
+                      crossval_method='kfold', K=5, max_crossval=np.inf,
+                      fit_kw=dict(lr=(1e-1, 1e-2), niter=(250, 500), progressbar=False),
+                      **model_kw):
+    """Performs cross-validation over number of components, warp regularization scale, and
```

Comment:
also it looks like warp_reg and time_reg have a weird expected format of functions that take a penalty amount as an argument, please doc!